### DISCUSSION TITLE
How to load modules that don't exist on file system?

### DISCUSSION THREAD

**Discussion by brundonsmith** [swc-project/swc#6927] [Q&A] [answered]
I'm attempting to use SWC directly as a Rust crate and build a tool on top of it

My tool generates its own JS files in memory, and I'd like to then pass them to SWC for transpiling and bundling before writing the final bundle to disk

It seems like what I'd want to do is create `FileName::Custom()`s for each in-memory module, however, looking at `SwcLoader` this goes down a path that leads it to always just spit out `module.exports = {}`:
```rust
fn load_with_handler(&self, handler: &Handler, name: &FileName) -> Result<ModuleData, Error> {
    tracing::debug!("JsLoader.load({})", name);
    let helpers = Helpers::new(false);

    if let FileName::Custom(id) = name {
        // Handle built-in modules
        if id.starts_with("node:") {
            let fm = self
                .compiler
                .cm
                .new_source_file(name.clone(), "".to_string());
            return Ok(ModuleData {
                fm,
                module: Module {
                    span: DUMMY_SP,
                    body: Default::default(),
                    shebang: Default::default(),
                },
                helpers: Default::default(),
            });
        // Handle disabled modules, eg when `browser` has a field
        // set to `false`
        } else {
            // TODO: When we know the calling context is ESM
            // TODO: switch to `export default {}`.
            let fm = self
                .compiler
                .cm
                .new_source_file(name.clone(), "module.exports = {}".to_string());

            let module = parse_file_as_module(
                &fm,
                Syntax::Es(Default::default()),
                Default::default(),
                None,
                &mut vec![],
            )
            .unwrap();
            return Ok(ModuleData {
                fm,
                module,
                helpers: Default::default(),
            });
        }
    }
```

Is this:
1. an unimplemented feature?
2. a misunderstanding on my part of what `FileName::Custom` represents? or,
3. do I need to implement a custom loader and not use `SwcLoader`?

**Comment by kdy1** [comment, marked as answer]
You mean you are using `swc_bundler`? It's not actively maintained in favor of turbopack

**Reply by brundonsmith** [reply]
Oh, I didn't realize that. Yes I'd started down the path of trying to get it working

Is turbopack available for use as a Rust library, on crates.io? A cursory glance there didn't really find anything

**Reply by brundonsmith** [reply]
Found a sorta-answer here: https://github.com/vercel/turbo/discussions/2370

I'll mark this question as answered, thanks
